File: frontend/src/api/rightsizing.ts
import axios from './axios'

export interface RightSizingRecommendation {
  resource_arn: string
  resource_name: string
  resource_type: string
  current_config: string
  recommended_config: string
  finding: string
  region?: string
  cpu_utilization?: number
  memory_utilization?: number
  performance_risk?: number
  estimated_monthly_savings: number
  savings_percentage?: number
  recommendation_source: string
}

export interface RightSizingRecommendationsResponse {
  profile_name: string
  total_recommendations: number
  total_monthly_savings: number
  recommendations_by_type: Record<string, number>
  recommendations: RightSizingRecommendation[]
}

export interface RightSizingSummary {
  profile_name: string
  total_ec2_recommendations: number
  total_ebs_recommendations: number
  total_lambda_recommendations: number
  total_asg_recommendations: number
  total_potential_savings: number
  overprovisioned_resources: number
  underprovisioned_resources: number
  optimized_resources: number
}

export const getRightSizingRecommendations = async (
  profileName: string,
  resourceTypes?: string
): Promise<RightSizingRecommendationsResponse> => {
  const response = await axios.get('/rightsizing/recommendations', {
    params: { profile_name: profileName, resource_types: resourceTypes }
  })
  return response.data
}

export const getRightSizingSummary = async (
  profileName: string
): Promise<RightSizingSummary> => {
  const response = await axios.get('/rightsizing/summary', {
    params: { profile_name: profileName }
  })
  return response.data
}

export const getTopSavingsOpportunities = async (
  profileName: string,
  limit: number = 10
): Promise<RightSizingRecommendation[]> => {
  const response = await axios.get('/rightsizing/top-opportunities', {
    params: { profile_name: profileName, limit }
  })
  return response.data
}

File: backend/app/schemas/rightsizing.py
from pydantic import BaseModel, Field
from typing import Optional, List


class RightSizingRecommendation(BaseModel):
    """Schema for a right-sizing recommendation"""
    resource_arn: str = Field(..., description="AWS resource ARN")
    resource_name: str = Field(..., description="Resource name/ID")
    resource_type: str = Field(..., description="Type: ec2_instance, ebs_volume, lambda_function, auto_scaling_group")
    current_config: str = Field(..., description="Current configuration (e.g., 't3.large', 'gp3 100GB')")
    recommended_config: str = Field(..., description="Recommended configuration")
    finding: str = Field(..., description="Finding: Underprovisioned, Overprovisioned, Optimized, etc.")
    region: Optional[str] = Field(None, description="AWS region where the resource is located")
    cpu_utilization: Optional[float] = Field(None, description="CPU utilization percentage")
    memory_utilization: Optional[float] = Field(None, description="Memory utilization percentage")
    performance_risk: Optional[float] = Field(None, description="Performance risk score (0-5)")
    estimated_monthly_savings: float = Field(..., description="Estimated monthly savings in USD")
    savings_percentage: Optional[float] = Field(None, description="Savings percentage")
    recommendation_source: str = Field(default="aws_compute_optimizer", description="Source of recommendation")


class RightSizingRecommendationsResponse(BaseModel):
    """Schema for right-sizing recommendations response"""
    profile_name: str
    total_recommendations: int
    total_monthly_savings: float
    recommendations_by_type: dict  # {resource_type: count}
    recommendations: List[RightSizingRecommendation]


class RightSizingSummary(BaseModel):
    """Schema for right-sizing summary"""
    profile_name: str
    total_ec2_recommendations: int
    total_ebs_recommendations: int
    total_lambda_recommendations: int
    total_asg_recommendations: int
    total_potential_savings: float
    overprovisioned_resources: int
    underprovisioned_resources: int
    optimized_resources: int
